Add  (test reports in json) and exit code to env before onpass/onfail

bendtherules opened the pull request:
Added:

1. ReportCollectorPlugin, which acts as pytest plugin to capture all test reports in a list with each test report being a dict, filled with `duration`, `failed`, `fspath`, `keywords`, `location`, `longrepr`, `nodeid`, `outcome`, `passed`, `sections`, `skipped`, `when` ; then convert them to json and store in env. variable `PYTEST_REPORTS`. Three reports of `setup`, `main test` and `teardown` is obtained for each actual test case.
2. Store pytest exit code in env as `PYTEST_EXIT_CODE`

Bug:
1. Env variable can be max 32767 bytes long, PYTEST_REPORTS may cross that.

bendtherules commented:
PR for #22

joeyespo commented on pytest_watch/watcher.py:
```diff
@@ -92,8 +94,8 @@ def run(self, summary=None):
             os.system(self.beforerun)
-        exit_code = subprocess.call(['py.test'] + self.args,
-                                    shell=(sys.platform == 'win32'))
+        exit_code = pytest.main(self.args, plugins=[ReportCollectorPlugin()])
```
Unfortunately, I don't think this will work. Try this for yourself, but last time I tried to use `pytest.main` instead of subprocess, code updates weren't reflected in the tests. I'm guessing it's because pytest is importing your tests and therefore application code, and so further calls to `pytest.main` uses the in-memory modules instead of re-importing.

Instead of attempting to bust cached modules, using subprocess guarantees that `py.test` is working with freshly imported modules as you make changes.

bendtherules replied:
Ok, I am taking another try with using subprocess, changing ReportCollectorPlugin into a  actual registered pytest plugin and using tempfile instead of env variable, whose path would then be set in env for onpass/onfail code.

joeyespo replied:
@bendtherules In light of something like `--runner`, I'm not sure adding `ReportCollectorPlugin` is worth the additional code. Unless we can figure out a way to have ptw to blindly pass results on to the callback executable or something.

I wouldn't mind adding some examples to the Readme.

bendtherules replied:
Ok.

In that case, can we have a external plugin set some value in pytest.config which would be passed on to the executable by ptw?

This way, ptw would have to not know about any plugins, just the value at `sessionexit`.

bendtherules commented:
Current Changes made :
1. Move plugin class to its own module, use it in setup.py pytest entry_point. This plugin needs a --report_collector_filename option with pytest where it stores its report
2. sample_onpass_onfail.py prints no. of passed/total cases and added to demonstrate use of onpass and onfail. Also, added it to console_scripts, but probably shouldnt be finally added to pytest-watch. Just a sample demo.

Run `ptw --onpass=sample_onpass_onfail --onfail=sample_onpass_onfail` to see it in action, it prints summary like `pass:x/x` after all tests

joeyespo commented:
RE: https://github.com/joeyespo/pytest-watch/issues/22#issuecomment-161147361 Let's continue the discussion here.

Passing the passed / failed count to `--onfail` seems good, but only if it doesn't require a lot of custom logic in ptw to do so.

My rationale is that both `--onfail` and `--runner` require custom end-user code. If passing along the passed / failed counts to `--onfail` requires a bunch of additional moving parts to the ptw internals (such as a custom plugin, a new console entry point, ENV variables, temp files, and a custom JSON contract that only contains partial result information), it doesn't seem worth it. Especially since it's only an additional few lines in the end-user `--runner` code to get the same information yourself.

Adding helpers to ptw is justifiable though because they can go a long way without adding overhead. Unlike internal code, an end-user script can extend the helpers to get more out of them for example.

Does that make sense?

joeyespo commented:
Another approach could be what @blueyed first suggested, reading the output of `py.test` (ex: scanning the last line of the output for `r'=+ (\d+) passed, (\d+) error in (\d+\d+) seconds =+'`, but only when `--onpass` or `--onfail` is present). This would give you the counts that you can pass as CLI arguments to the end-user script for convenience without much overhead.

Although even this might be overcomplicating things given a `--runner` + good helpers.

bendtherules commented:
The regex options sounds very bad to me. I dont think its dependable enough considering that you can pass a bunch of flags to pytest to change the output, even user code can change it.

I am not sure about the philosophy of ptw, which I think will ultimately help us choose the soln: Is it supposed to "know" about pytest or, it is supposed to be a python on-change test runner, that just "supports" pytest for now but should ideally be able to handle other test frameworks?

If it is the second, then runner seems to be in the right direction.

If the first, then I think we should re-evaluate it -
We are trying to let the user provide their own plugin for providing test result data, but pass it to the onpass/onfail test cases ourself. So, currently we allow a custom runner - but the problem is that we're leaving off the task of actually calling `py.test` to the runner, which is the most important call and not something I think a plugin needs to change. But the runner is actually there to facilitate the allow of custom plugin, right?

So, instead of bothering about a runner, which is arguably a plugin to a plugin(i.e. ptw) of pytest - we should just ask user to use their own pytest custom plugin and implement a scraper plugin within ptw, which will read a pre-decided value off pytest.config in the end and provide it as param to onpass/onfail.

That way, there is no json contract between ptw and onfail executable, it is between plugin and executable - both of which is user-defined. Also NO  new console entry point, ENV variables or temp files - only a Pytest plugin within ptw, which I dont think is any reason to avoid.

Does that make sense?

joeyespo commented:
@bendtherules Agreed that regex still isn't optimal.

Pytest-watch does indeed know about pytest, but adding `--runner` doesn't exactly violate the "knows about pytest" concern any more than subclassing a class to customize behavior goes against the semantics of the base class. (Well in both cases, you obviously *can*. But that violates [the Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle), which is usually not a good idea. Similarly in ptw, if you replace pytest with something else it's up to you to deal with the consequences of that decision.)

I'm not completely against having a plugin for pytest or providing some extra information to `--onpass` or `--onfail`, but I'm cautious about complecting the internals if there's another way to get the information you're after that doesn't do that.

joeyespo commented:
Closing this in favor of the `--runner` approach.
